Fixes to Merging algorithm

padupr opened the pull request:
Last minute fixes.

Commit:
Fixes to merging algorithm to compensate xml formats limitations

Also tests

Commit:
Fix

Commit:
Fix

Commit:
Workaround for getChildren, relationship copy for non strict fixed

Commit:
Stuff

Commit:
Eval version

vulder committed:
Merge branch 'vara-dev' into f-mergingFixes

vulder commented:
@padupr What's the current status of this PR?

vulder committed:
Update include/vara/Feature/FeatureTreeNode.h

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Commit:
Merge branch 'vara-dev' into f-mergingFixes

# Conflicts:
#	include/vara/Feature/FeatureTreeNode.h
#	unittests/resources/CMakeLists.txt

Commit:
Revert merge error

Commit:
Remove old testcase, reproduced in other tests

s9latimm commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -286,6 +286,14 @@ class FeatureModel {
+    Constraints.erase(
+        std::find_if(Constraints.begin(), Constraints.end(),
+                     [C](const std::unique_ptr<Constraint> &UniC) {
+                       return UniC.get() == C;
```
Here we compare two pointers (instead of the underlying constraints) -- maybe we should replace this in the future with tree based constraint comparison.

vulder replied:
If we don't have them yet, a TODO that refers to the issue would be good.

s9latimm replied:
https://github.com/se-sic/VaRA/issues/701

s9latimm commented on lib/Feature/FeatureModelWriter.cpp:
```diff
@@ -25,8 +25,9 @@ static constexpr char ENCODING[] = "UTF-8";
   int RC;
   std::unique_ptr<xmlTextWriter, void (*)(xmlTextWriterPtr)> Writer(
-      xmlNewTextWriterFilename(Path.data(), 0), &xmlFreeTextWriter);
+      xmlNewTextWriterFilename(Path.c_str(), 0), &xmlFreeTextWriter);
```
Does this change anything?

https://stackoverflow.com/questions/194634/string-c-str-vs-data

vulder replied:
should not matter anymore

s9latimm commented on unittests/Feature/FeatureModelParser.cpp:
```diff
@@ -29,17 +29,6 @@ TEST(FeatureModelParser, onlyChildren) {
   EXPECT_TRUE(P.buildFeatureModel());
 }
 
-TEST(FeatureModelParser, onlyParents) {
-  auto FS = llvm::MemoryBuffer::getFileAsStream(
-      getTestResource("test_only_parents.xml"));
-  assert(FS);
-
-  auto P = FeatureModelXmlParser(FS.get()->getBuffer().str());
-
-  EXPECT_TRUE(P.verifyFeatureModel());
-  EXPECT_TRUE(P.buildFeatureModel());
-}
-
```
Why is this test disabled? The goal was to test solely the behavior of the parser.

padupr replied:
Not sure why I removed that. Reverted.

s9latimm commented on lib/Feature/FeatureModelTransaction.cpp:
```diff
@@ -147,9 +178,9 @@ mergeFeatureModels(FeatureModel &FM1, FeatureModel &FM2, bool Strict) {
-  if (F1.getParent()->getKind() != F2.getParent()->getKind()) {
-    return false;
-  }
+  //  if (F1.getParent()->getKind() != F2.getParent()->getKind()) {
+  //    return false;
+  //  }
```
Why is this commented out?

padupr replied:
I needed to allow merging of subtrees if the relationship is not present in both, so the parent node is no longer required to be equal.

vulder replied:
So can we remove this or do we need a TODO to implement something else?

padupr replied:
Removed it, no need to implement something else.

s9latimm commented on unittests/Feature/FeatureModelTransaction.cpp:
```diff
@@ -394,6 +397,228 @@ TEST_F(FeatureModelMergeTransactionTest, AcceptNonStrictDifferenceKind) {
+  auto FMMerged = mergeFeatureModels(*FM2, *FM1);
+  ASSERT_TRUE(FMMerged);
+
+  ASSERT_TRUE(FMMerged->getFeature("a"));
+  ASSERT_TRUE(FMMerged->getFeature("aa"));
+  EXPECT_EQ(FMMerged->getFeature("aa")->getParent()->getKind(),
+            FeatureTreeNode::NodeKind::NK_RELATIONSHIP);
+  ASSERT_TRUE(FMMerged->getFeature("ab"));
+  EXPECT_EQ(FMMerged->getFeature("ab")->getParent()->getKind(),
+            FeatureTreeNode::NodeKind::NK_RELATIONSHIP);
```
This may be refactored with using `TearDown` in a base test.

s9latimm reviewed (changes requested):
Over all LGTM, just some remarks / questions unrelated to the actual algorithm.

vulder commented on unittests/Feature/FeatureModelTransaction.cpp:
```diff
@@ -1,9 +1,12 @@
 #include "vara/Feature/FeatureModelTransaction.h"
 #include "vara/Feature/FeatureModelBuilder.h"
 
+#include "UnittestHelper.h"
 #include "gtest/gtest.h"
 
+#include <llvm/Support/MemoryBuffer.h>
 #include <memory>
+#include <vara/Feature/FeatureModelParser.h>
```
```suggestion
#include "vara/Feature/FeatureModelTransaction.h"
#include "vara/Feature/FeatureModelBuilder.h"
#include "vara/Feature/FeatureModelParser.h"

#include "UnittestHelper.h"

#include "gtest/gtest.h"

#include "llvm/Support/MemoryBuffer.h"

#include <memory>

```

vulder commented on unittests/Feature/FeatureModel.cpp:
```diff
@@ -224,6 +224,34 @@ TEST_F(FeatureModelTest, dfs) {
 //                    FeatureModelConsistencyChecker Tests
```
Copy paste error, I think you wanted to put the test name there.

vulder commented on lib/Feature/FeatureModelTransaction.cpp:
```diff
@@ -7,10 +7,15 @@ namespace vara::feature {
 bool mergeSubtree(FeatureModelCopyTransaction &Trans, FeatureModel const &FM,
                   Feature &F, bool Strict);
 
+bool mergeSubtree(FeatureModelCopyTransaction &Trans, FeatureModel const &FM,
+                  Relationship &R, bool Strict);
```
I'm not 100% sure what this method does, but maybe the name is just confusing.
Does this merge feature subtree that are related to the Relationship or does it merge relationships?

padupr replied:
It's unused so I removed it. Originally, planned to overload the function to handle Relationships differently during merging.

vulder replied:
So, if you don't need that anymore a updated name should suffice

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -803,14 +861,19 @@ class FeatureModelCopyTransactionBase {
-    if (!FM) {
-      return nullptr;
-    }
+    assert(FM && "FeatureModel is null.");
```
Do we want to just assert here or return by default a nullptr when we are called with no FM?
@s9latimm What do you think?

padupr replied:
I had a look at other Transactions. It's handled in three ways: check and return nullptr, assert, no check. I think this should be solved after error codes are implemented.

vulder replied:
Yes, error codes should make this case clear, so let's keep it that way.

vulder commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -365,6 +414,15 @@ class RemoveFeatureFromModel : public FeatureModelModification {
       }
     }
 
+    while (!F->constraints().empty()) {
```
If I understand this correctly, we go to the most upper constraint and delete it. You can we prevent that this delete much more than was intended? Not sure if a edge case can happen here, where we remove to much.

padupr replied:
This code should: 

1. Pick a constraint of the feature
2. Traverse upwards to get the constraint's expression tree root
3. Remove the whole Constraint from the model, this will also remove it from all other related features
4. Continue with the next constraint

We might want to offer a choice on the behavior in the future, e.g. error if constraints are present, try to resolve the deletion based on rules, remove all present constraints on the feature entirely (current solution).

vulder replied:
Ok, sounds ok to me. Could you please create an issue for that an reference it here with an TODO, so we come back to this later :)

Commit:
Feedback

padupr committed:
Update unittests/Feature/FeatureModelTransaction.cpp

Co-authored-by: Florian Sattler <vuld3r@gmail.com>

padupr committed:
Update unittests/Feature/FeatureModelTransaction.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

vulder commented:
What is the current state of this PR? @padupr could you please, if you find time, go over the open remarks and close them if they are resolved.

Commit:
Feedback

vulder committed:
Merge branch 'vara-dev' into f-mergingFixes

vulder commented:
@padupr I pulled in the current fixes on dev. Otherwise, we need to remove the issues reported from the CI.

Commit:
Revert changes to bindings

Commit:
Revert all changes to bindings

vulder committed:
Merge branch 'vara-dev' into f-mergingFixes

vulder committed:
Fixes error

vulder commented:
@s9latimm I updated the PR to our ne dev state. Can you please double check this PR, especially the TODO part regarding error handling. What should we do in this case? Do we need a specific error here?

s9latimm commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -326,6 +342,46 @@ class AddFeatureToModel : public FeatureModelModification {
+  }
+
+  Result<FTErrorCode> operator()(FeatureModel &FM) {
+    // TODO: Any error handling needed here?
+    assert(RemoveConstraint.getParent() == nullptr &&
+           "Subtree deletion is not supported");
```
We should talk about this today -- as far as I understand, uncoupling currently only works for this case.

vulder committed:
Adds error

vulder committed:
Merge branch 'vara-dev' into f-mergingFixes

vulder committed:
Merge branch 'vara-dev' into f-mergingFixes

vulder committed:
Adapts constraint removal to new feature-constraint handles

vulder committed:
Updates external dependencies

vulder committed:
Merge branch 'update-dependencies' into f-mergingFixes

vulder committed:
Removes unnecessary parent lookup

It currently seems that removing the constraint from a parent seems
wrong. However, there might be a case we miss here. So, should we
encounter a breakages from that, we should revisit the impl. here.

vulder commented:
- [x] Depends on #133

vulder commented:
@boehmseb @Kallistos please have a look. I tried to rebase and clean up the work from Pascal but an additional feedback would be good.

vulder committed:
Merge branch 'vara-dev' into f-mergingFixes

boehmseb commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -254,6 +258,14 @@ class FeatureModelModification {
     return FM.addConstraint(std::move(Constraint));
   }
 
+  static void removeConstraint(FeatureModel &FM, Constraint *R) {
```
Why `Consraint *R` here and `Constraint *C` below?
Do they have different meaning, or should they be named the same?

vulder replied:
Don't know, probably an over site from @padupr , I'll rename this to C.

boehmseb commented on include/vara/Feature/FeatureModel.h:
```diff
@@ -399,6 +399,48 @@ class FeatureModel {
+                         return UniC->constraint() == C;
+                       });
+      if (MCIt != MixedConstraints.end()) {
+
```
```suggestion
```

boehmseb commented on include/vara/Feature/FeatureModelTransaction.h:
```diff
@@ -310,8 +322,12 @@ class AddFeatureToModel : public FeatureModelModification {
-      setParent(*InsertedFeature, *Parent);
-      addEdge(*Parent, *InsertedFeature);
+      FeatureTreeNode *ParentNode = Parent;
+      if (!Parent->getChildren<Relationship>(1).empty()) {
+        ParentNode = *(Parent->getChildren<Relationship>(1).begin());
+      }
```
Just to make sure I understand this correctly:
This is required because we model (xor/or) relationships as separate nodes so the relationship node has to be set as parent if one exists, correct?

This assumes that a Feature can have exactly one (or no) relationship as child and if it has, there are no features as direct children allowed. Is this assumption validated anywhere?
I'm asking because this might become an issue when integrating UVL because UVL does not have these constraints, i.e., a feature can have multiple types of relationships as children at once (I'm not sure how about multiple instances of the same type, e.g., multiple or groups).

vulder replied:
I'm not sure, if it's currently not possible to have multiple ones, we probably can adapt the API to make it viable.

boehmseb committed:
Merge branch 'vara-dev' into f-mergingFixes

vulder committed:
fix: Param name

vulder committed:
Update include/vara/Feature/FeatureModel.h

Co-authored-by: Sebastian Böhm <boehmseb@cs.uni-saarland.de>
